### PR TITLE
[kube-prometheus-stack] use of helm tpl for ingress annotations

### PR DESCRIPTION
**Pull request by richardtief**
#### What this PR does / why we need it

Use the Helm tpl function for ingress annotations to avoid setting already defined values again. 

e.g.
```yaml
annotations:
  nginx.ingress.kubernetes.io/auth-tls-secret: "{{ $.Release.Namespace }}/secret-name"
```
ref: https://kubernetes.github.io/ingress-nginx/user-guide/nginx-configuration/annotations/#client-certificate-authentication

This change is backwards compatible and will not break anything.

#### Which issue this PR fixes

*(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*

- fixes #

#### Checklist

- [x] [DCO](https://github.com/prometheus-community/helm-charts/blob/main/CONTRIBUTING.md#sign-off-your-work) signed
- [x] Chart Version bumped
- [x] Title of the PR starts with chart name (e.g. `[prometheus-couchdb-exporter]`)

@jkroepke @gianrubio @gkarthiks @GMartinez-Sisti @scottrigby

### CODE REVIEW

**Comment by jkroepke** [comment]
Could you please try this one:

```
annotations:
  nginx.ingress.kubernetes.io/auth-tls-secret: "/secret-name"
```

If namespace is empty, the ns of the ingress object is used.

https://github.com/kubernetes/ingress-nginx/blob/4a680c8b200cb865f8698ce0aef4123bfff32f3e/internal/ingress/annotations/authtls/main.go#L157-L159

**Comment by richardtief** [comment]
Thanks for the quick reply and the hint. Actually, I would also like to use this for the secret name that contains the namespace. 

```yaml
annotations:
  nginx.ingress.kubernetes.io/auth-tls-secret: "{{ $.Release.Namespace }}/{{ $.Release.Namespace }}-secret-prefix"
```

The tpl function is widely used in charts, so that shouldn't hurt, no?

File: charts/kube-prometheus-stack/templates/alertmanager/ingress.yaml [hunk 1]
```diff
@@ -14,7 +14,7 @@ metadata:
   namespace: {{ template "kube-prometheus-stack.namespace" . }}
 {{- if .Values.alertmanager.ingress.annotations }}
   annotations:
-{{ toYaml .Values.alertmanager.ingress.annotations | indent 4 }}
+{{ tpl (toYaml .Values.alertmanager.ingress.annotations) . | indent 4 }}
```
Comment (jkroepke) [inline charts/kube-prometheus-stack/templates/alertmanager/ingress.yaml:17]:
```suggestion
    {{- tpl (toYaml .Values.alertmanager.ingress.annotations) . | nindent 4 }}
```

File: charts/kube-prometheus-stack/templates/alertmanager/ingressperreplica.yaml [hunk 2]
```diff
@@ -25,7 +25,7 @@ items:
       {{- end }}
       {{- if $ingressValues.annotations }}
       annotations:
-{{ toYaml $ingressValues.annotations | indent 8 }}
+{{ tpl (toYaml $ingressValues.annotations) . | indent 8 }}
```
Comment (jkroepke) [inline charts/kube-prometheus-stack/templates/alertmanager/ingressperreplica.yaml:28]:
```suggestion
        {{- tpl (toYaml $ingressValues.annotations) . | nindent 8 }}
```

File: charts/kube-prometheus-stack/templates/prometheus/ingress.yaml [hunk 3]
```diff
@@ -11,7 +11,7 @@ kind: Ingress
 metadata:
 {{- if .Values.prometheus.ingress.annotations }}
   annotations:
-{{ toYaml .Values.prometheus.ingress.annotations | indent 4 }}
+{{ tpl (toYaml .Values.prometheus.ingress.annotations) . | indent 4 }}
```
Comment (jkroepke) [inline charts/kube-prometheus-stack/templates/prometheus/ingress.yaml:14]:
```suggestion
    {{- tpl (toYaml .Values.prometheus.ingress.annotations) . | nindent 4 }}
```

File: charts/kube-prometheus-stack/templates/prometheus/ingressThanosSidecar.yaml [hunk 4]
```diff
@@ -11,7 +11,7 @@ kind: Ingress
 metadata:
 {{- if .Values.prometheus.thanosIngress.annotations }}
   annotations:
-{{ toYaml .Values.prometheus.thanosIngress.annotations | indent 4 }}
+{{ tpl (toYaml .Values.prometheus.thanosIngress.annotations) . | indent 4 }}
```
Comment (jkroepke) [inline charts/kube-prometheus-stack/templates/prometheus/ingressThanosSidecar.yaml:14]:
```suggestion
    {{- tpl (toYaml .Values.prometheus.thanosIngress.annotations) . | nindent 4 }}
```

File: charts/kube-prometheus-stack/templates/prometheus/ingressperreplica.yaml [hunk 5]
```diff
@@ -25,7 +25,7 @@ items:
       {{- end }}
       {{- if $ingressValues.annotations }}
       annotations:
-{{ toYaml $ingressValues.annotations | indent 8 }}
+{{ tpl (toYaml $ingressValues.annotations) . | indent 8 }}
```
Comment (jkroepke) [inline charts/kube-prometheus-stack/templates/prometheus/ingressperreplica.yaml:28]:
```suggestion
        {{- tpl (toYaml $ingressValues.annotations) . | nindent 8 }}
```

File: charts/kube-prometheus-stack/templates/thanos-ruler/ingress.yaml [hunk 6]
```diff
@@ -13,7 +13,7 @@ metadata:
   namespace: {{ template "kube-prometheus-stack.namespace" . }}
 {{- if .Values.thanosRuler.ingress.annotations }}
   annotations:
-{{ toYaml .Values.thanosRuler.ingress.annotations | indent 4 }}
+{{ tpl (toYaml .Values.thanosRuler.ingress.annotations) . | indent 4 }}
```
Comment (jkroepke) [inline charts/kube-prometheus-stack/templates/thanos-ruler/ingress.yaml:16]:
```suggestion
    {{- tpl (toYaml .Values.thanosRuler.ingress.annotations) . | nindent 4 }}
```

**Review by jkroepke — COMMENTED** [review]
Since Helm 3.14, tpl functions doent hurt anymore..

I'm generally fine with this, but I would like to pull a second review

**Comment by jkroepke** [comment]
@richardtief please do a rebase. I guess its good to go.

**Comment by richardtief** [comment]
Thanks @jkroepke. Rebase is done.

File: charts/kube-prometheus-stack/Chart.yaml [hunk 7]
```diff
@@ -23,7 +23,7 @@ name: kube-prometheus-stack
 sources:
   - https://github.com/prometheus-community/helm-charts
   - https://github.com/prometheus-operator/kube-prometheus
-version: 56.7.0
+version: 56.7.1
```
Comment (jkroepke) [inline charts/kube-prometheus-stack/Chart.yaml:26]:
```suggestion
version: 56.8.0
```

**Review by jkroepke — COMMENTED** [review]
LGTM

Since its a new feature, please bump the minor version.

**Review by jkroepke — APPROVED** [review]
LGTM 👍

**Comment by aaronriedel** [comment]
It looks like this breaks current helm values when used with helm 3.13 or before. Helm 3.14 is out for only about 1 month now. Maybe this could have been stated in the changelogs.

**Comment by jkroepke** [comment]
Could you provide an example what is not working with helm 3.13?

In case, you refer my comment

> Since Helm 3.14, tpl functions doent hurt anymore..

Everything works fine with 3.13. In 3.14 , the performance issues with the `tpl` are fixed. 

Before 3.14, using more `tpl` function just slows the helm template.

**Comment by aaronriedel** [comment]
@jkroepke I have described it here: https://github.com/prometheus-community/helm-charts/issues/4269

After I updated my argocd it uses helm 3.14 and it works fine now. With helm 3.13 my traefik annotation was throwing an error.

**Comment by jkroepke** [comment]
@aaronriedel I looked into this and this is not intended. I looked into we have a bug here.

It could possible that helm 3.14 has a fix here, but the tpl usage in ingressperreplica is broken

**Comment by aaronriedel** [comment]
Okay, thank you so much for looking into this :) 
Maybe other users will bump into this eventually, because not everyone keeps their helm or argocd or whatever up to date this fast.

**Comment by richardtief** [comment]
Bummer! There is another one for the alertmanager. https://github.com/prometheus-community/helm-charts/blob/f71c7dab23d83e8646203809a6999fe7b95f2f74/charts/kube-prometheus-stack/templates/alertmanager/ingressperreplica.yaml#L28

**Comment by jkroepke** [comment]
https://github.com/prometheus-community/helm-charts/pull/4271

**Pull request merged**